### ISSUE TITLE
Allowing `EventLoop::run` to return - Discussion / Proposal for a new `EventLoop::run`

### ISSUE THREAD

**Issue by mitchmindtree** [RustAudio/cpal#245]
This is a more general follow-up to #236 (cc @quadrupleslap) that aims to explore a few more options.

## The Problem

Currently calling `EventLoop::run` takes ownership over the thread ensuring that it never returns. This results in a couple of issues:

- Does not allow for handling any errors that might occur.
- Requires that users spawn a whole thread just to run the event loop, even if the underlying OS audio API does not block.
- Makes it difficult to manually clean up resources (like exclusive audio driver handles in ASIO cc @freesig) when the program ends.

## Potential Solutions

There are a couple of ways we could signal to the `EventLoop` to break:

1. Add an `EventLoop` method.
2. A return type within the user callback.

An `EventLoop` method allows for signaling from other threads (useful if controlling the `EventLoop` from a GUI for example) and from within the audio thread itself (though whether or not the `EventLoop` should even be accessible within the callback is questionable, as it should not be possible to call `run` again recursively).

A "control flow" style return type within the user callback allows for signaling without the need for access to the `EventLoop` within the callback itself. By using a `Result`, we can also more easily distinguish between "successful" loop breaks and loop breaks that are due to some kind of error.

Perhaps a hybrid between the two could be suitable, e.g. an `EventLoop` method for inter-thread signaling and a callback return type to allow for signaling from the audio thread?

Perhaps the new `EventLoop::run` signature could look something like this:

```rust
    pub fn run<F>(&self, mut callback: F) -> Result<(), RunError>
    where
        F: FnMut(StreamId, StreamData) -> Result<ControlFlow, Box<Error>> + Send,
```

where the `ControlFlow` and `RunError` look something like this:

```rust
pub enum ControlFlow {
    Continue,
    Break,
}

pub enum RunError {
    /// An error produced by the user callback.
    Callback(Box<Error>),
    /// The device was unexpectedly unavailable (e.g. disconnected).
    DeviceUnavailable(StreamId),
    // etc,
}
```

This won't be a trivial fix as it may require some control flow acrobatics within some of the backends. For example, on initial investigation into updating the ALSA backend it is obvious that a few choices have been made under the assumption that the `run` function will loop forever. One of these is that streams are added, removed, played and paused via a queue of `Command`s, which are only processed at the beginning of each `run` loop pass. In the case that commands are queued during a call to the user callback, should they be processed prior to breaking if the user callback indicates breaking should occur?

Further, we should probably make sure to "stop" all of the streams that are currently playing if an indication has been made to break from the loop. Otherwise, platforms with non-blocking audio processing (like macOS) will probably just keep playing even though the `EventLoop` is no longer "running".

Something worth keeping in mind that is related to this work is whether or not we should eventually allow for a way of distinguishing between blocking and non-blocking loops. I know this became an important goal for supporting low-powered devices in the portaudio dev, but I don't have any reference links on me at the moment. Perhaps it's worth saving this discussion for when `futures` lands in `std`, as it might be a suitable abstraction.

**Comment by mitchmindtree** [comment]
## Handling active `Stream`s when breaking from `run`

Before implementing this, a decision has to be made about what happens to `Stream`s that are playing when the user breaks from the loop.

Intuitively, I would expect for all streams to be "stop"ped so that they stop attempting to call the user callback for more samples. The first thought that comes to mind is that we could call `pause_stream` on all of the existing, playing streams before returning. There are a few issues that come to mind with this:

- If the user were to call `run` again, should they expect that the streams that were playing when `run` last ended automatically start again? Or should they have to first manually `play_stream` each of those streams again? In the first case, the backend would have to track which streams were stopped due to `run` returning and then restart those behind the scenes when `run` is called again.
- What about devices that don't support hardware pausing? #197. Should we swap the user callback with one that generates silence?
- What does it mean to `play_stream` or `pause_stream` an `EventLoop` that isn't currently running? In the ALSA backend, calling play/pause stream enqueues "commands" that the event loop will only begin processing once `run` is called. Perhaps this should be the enforced behaviour for all backends?

**Comment by xSke** [comment]
I was briefly looking into making cpal work with the `wasm32-unknown-unknown` target. It seems feasible, but cpal's emscripten backend relies on a call to `stdweb::event_loop()` to essentially "block" the thread. That call's a no-op using `cargo-web` or the `wasm32-unknown-unknown` target, and doesn't even return `!` on newer versions of `stdweb`, so this is essentially impossible to "make work".

**Comment by mitchmindtree** [comment]
Thanks a lot for the report @xSke, very interesting to hear and good to know! I had a feeling the original reason for not returning was due to a limitation along these lines. Right now there's quite a bit of interest in removing the `EventLoop` altogether (see #278), but I'd like to make sure we have a solid plan for web before we go ahead with it, if at all.

**Comment by mitchmindtree** [comment]
Solved via #354
